core/chains/evm/gas/rollups/op_l1_oracle.go

```diff
@@ -0,0 +1,490 @@
+package rollups
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/rpc"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+	"github.com/smartcontractkit/chainlink-common/pkg/utils"
+
+	gethtypes "github.com/ethereum/go-ethereum/core/types"
+
+	"github.com/smartcontractkit/chainlink/v2/common/client"
+	"github.com/smartcontractkit/chainlink/v2/common/config"
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/assets"
+	evmclient "github.com/smartcontractkit/chainlink/v2/core/chains/evm/client"
+)
+
+// Reads L2-specific precompiles and caches the l1GasPrice set by the L2.
+type optimismL1Oracle struct {
+	services.StateMachine
+	client     l1OracleClient
+	pollPeriod time.Duration
+	logger     logger.SugaredLogger
+	chainType  config.ChainType
+
+	l1GasPriceAddress   string
+	gasPriceMethod      string
+	l1GasPriceMethodAbi abi.ABI
+	l1GasPriceMu        sync.RWMutex
+	l1GasPrice          priceEntry
+
+	l1GasCostAddress   string
+	gasCostMethod      string
+	l1GasCostMethodAbi abi.ABI
+
+	priceReader daPriceReader
+
+	chInitialised chan struct{}
+	chStop        services.StopChan
+	chDone        chan struct{}
+}
+
+type opStackGasPriceReader struct {
+	client l1OracleClient
+	logger logger.SugaredLogger
+
+	oracleAddress      common.Address
+	isEcotoneMethodAbi abi.ABI
+
+	l1BaseFeeCalldata    []byte
+	isEcotoneCalldata    []byte
+	getL1GasUsedCalldata []byte
+	getL1FeeCalldata     []byte
+
+	isEcotone        bool
+	isEcotoneCheckTs int64
+}
```

Comment:
I think we can move all of the logic and fields from the `opStackGasPriceReader` to `optimismL1Oracle`. Think it'd improve readability if everything was consolidated into a single struct. The reason this was created as a separate struct before was because it was hard to incorporate in the generic l1 oracle. Now that we have an OP specific oracle we can just merge these. You'll need to consolidate the builder methods below into one too.